### PR TITLE
API Review Changes

### PR DESCRIPTION
**Pull request by conniey**
* Adds convenience overloads and tests for:
  * addSetting
  * setSetting
  * updateSetting
* After reading the spec, ConfigurationSetting.value is a mandatory field, so I've updated the docs to say that. 
* Improves javadocs based on feedback
* Moves `ClientConfigurationBuilder` into its own class.
   * It appears a lot cleaner this way.
* Changes EnumSet<T> into varargs for SettingSelector.fields(EnumSet<T>). 
   * It makes it a lot easier to call the method because you don't have to create an EnumSet when you only use it once. 
* Reduce the number of custom policies (deleted RequestIdPolicy)
* Reduce the work of ConfigurationCredentialsPolicy
  * Using AddHeadersPolicy and AddDatePolicy

### CODE REVIEW

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java [hunk 1]
```diff
@@ -26,22 +19,20 @@
...
-import java.util.EnumSet;
-import java.util.List;
+import java.util.Arrays;
 import java.util.Locale;
 import java.util.Objects;
 import java.util.stream.Collectors;
 
 /**
- * Client that contains all the operations for KeyValues in Azure Configuration Store.
+ * Client that contains all the operations for {@link ConfigurationSetting ConfigurationSettings} in Azure Configuration Store.
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: I'm feeling like we need to have a max line length rule, at least for JavaDoc, in CheckStyle.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java [hunk 2]
```diff
@@ -26,22 +19,20 @@
...
+ *
+ * @see ConfigurationClientBuilder
+ * @see ConfigurationClientCredentials
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: Love it!

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java [hunk 3]
```diff
@@ -26,22 +19,20 @@
...
-import java.util.List;
+import java.util.Arrays;
 import java.util.Locale;
 import java.util.Objects;
 import java.util.stream.Collectors;
 
 /**
- * Client that contains all the operations for KeyValues in Azure Configuration Store.
+ * Client that contains all the operations for {@link ConfigurationSetting ConfigurationSettings} in Azure Configuration Store.
+ *
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: In the fullness of time this client class is a prime location for a few paragraphs of useful text, code samples, etc. Add that to your TODO list (or to Alan's)

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java [hunk 4]
```diff
@@ -75,60 +79,87 @@ public static Builder builder() {
      * The label value for the ConfigurationSetting is optional.
      *
      * @param setting The setting to add to the configuration service.
-     * @return ConfigurationSetting that was created or updated.
+     * @return ConfigurationSetting that was created.
      * @throws NullPointerException If {@code setting} is {@code null}.
-     * @throws IllegalArgumentException If {@link ConfigurationSetting#key()} is {@code null} or an empty string.
+     * @throws IllegalArgumentException If {@link ConfigurationSetting#key() key} or {@link ConfigurationSetting#value() value}
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: Good linking :-)

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java [hunk 5]
```diff
@@ -75,60 +79,87 @@ public static Builder builder() {
...
+     * are {@code null} or an empty string.
      * @throws ServiceRequestException If a ConfigurationSetting with the same key and label exists.
      */
     public Mono<Response<ConfigurationSetting>> addSetting(ConfigurationSetting setting) {
-        ConfigurationSetting result = validateSetting(setting);
-
+        ConfigurationSetting result = validateSetting(setting, true);
         return service.setKey(serviceEndpoint, result.key(), result.label(), result, null, getETagValue(ETAG_ANY));
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: I get a little tripped up by the fact that addSetting and setSetting methods both call service.setKey. It's probably right, but it causes confusion when I'm skimming. Do you think a clarifying internal comment might be useful for future developers?

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java [hunk 6]
```diff
@@ -75,60 +79,87 @@ public static Builder builder() {
...
      * @param setting The configuration setting to create or update.
      * @return ConfigurationSetting that was created or updated.
      * @throws NullPointerException If {@code setting} is {@code null}.
-     * @throws IllegalArgumentException If {@link ConfigurationSetting#key()} is {@code null} or an empty string.
-     * @throws ServiceRequestException If the {@link ConfigurationSetting#etag()} was specified, is not
-     * {@link ConfigurationClient#ETAG_ANY}, and the current configuration
-     * value's etag does not match.
+     * @throws IllegalArgumentException If {@link ConfigurationSetting#key() key} or {@link ConfigurationSetting#value() value}
+     * are {@code null} or an empty string.
+     * @throws ServiceRequestException If the {@link ConfigurationSetting#etag() etag} was specified, is not
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: Something for you and / or Alan to put on the TODO list is to explore better azconfig exception types....when the work @anuchandy is working on enables it.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java [hunk 7]
```diff
@@ -75,60 +79,87 @@ public static Builder builder() {
...
+        ConfigurationSetting result = validateSetting(setting, true);
 
         return service.setKey(serviceEndpoint, result.key(), result.label(), result, getETagValue(result.etag()), null);
     }
 
+    /**
+     * Updates an existing configuration value in the service with the given key. The setting must already exist.
+     *
+     * <p>
+     * Partial updates are not supported.
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: Weird JavaDoc for two reasons: missing a closing </p> element (more recent JavaDoc versions enforce HTML5 compliance), but also, it's weird to have a one-sentence paragraph. I would just add this to the end of the previous paragraph?

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java [hunk 8]
```diff
@@ -75,60 +79,87 @@ public static Builder builder() {
...
+     *
+     * @param key The key for the configuration setting to update.
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]:
```suggestion
     * @param key The key of the configuration setting to update.
```

Comment (JonathanGiles) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: I'm not sure if it should be 'for' or 'of', or if I should just not care. Consider it here and elsewhere, and do whatever you feel is better English.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java [hunk 9]
```diff
@@ -75,60 +79,87 @@ public static Builder builder() {
...
+    public Mono<Response<ConfigurationSetting>> updateSetting(String key, String value) {
+        return updateSetting(new ConfigurationSetting().key(key).value(value));
+    }
+
     /**
      * Updates an existing configuration value in the service. The setting must already exist.
      *
      * <p>
-     * The label value for the ConfigurationSetting is optional.
+     * Partial updates are not supported.
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: FIXME

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java [hunk 10]
```diff
@@ -75,60 +79,87 @@ public static Builder builder() {
...
      *
      * <p>
-     * If {@link ConfigurationSetting#etag()} is specified, the configuration value is only updated if it matches.
+     * If {@link ConfigurationSetting#etag() etag} is specified, the configuration value is only updated if it matches.
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: You're on a paragraph frenzy. I recommend generating the HTML output and ensuring you like what you see. I suspect it'll look a little sparse.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java [hunk 11]
```diff
@@ -240,14 +271,14 @@ public static Builder builder() {
...
-        ConfigurationSetting result = validateSetting(setting);
+        ConfigurationSetting result = validateSetting(setting, false);
 
         return service.unlockKeyValue(serviceEndpoint, result.key(), result.label(), null, null);
     }
 
     /**
      * Fetches the configuration settings that match the {@code options}. If {@code options} is {@code null}, then all the
-     * {@link ConfigurationSetting}s are fetched in their current state with default fields.
+     * {@link ConfigurationSetting configuration settings} are fetched in their current state.
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: "in their current state" is a bit hard to parse for someone unfamiliar with the service. Does it need a clarifying sentence?

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClientBuilder.java [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+/**
+ * Provides configuration options for instances of {@link ConfigurationClient}.
+ *
+ * @see ConfigurationClient
+ * @see ConfigurationClientCredentials
+ */
+public final class ConfigurationClientBuilder {
+    //TODO: Fix with https://github.com/Azure/azure-sdk-for-java/issues/3141
+    static final String SDK_NAME = "Azure-Application-Configuration";
+    static final String SDK_VERSION = "1.0.0-SNAPSHOT";
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClientBuilder.java]: It might be nice to extract this out into a package-private class, because we will want to centralise this for the sync client too. I would like to have a standard location for this for all client libraries, e.g. some `AzureConfig` or whatever class in the root package.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java (hunk 1 above)
Comment (conniey) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: Hmm.. I have my line rules at 120 characters. 🤔 We can enforce one in: .editorconfig if you want. https://github.com/editorconfig/editorconfig/wiki/EditorConfig-Properties#max_line_length

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java (hunk 8 above)
Comment (conniey) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: That's a great question. Ahaha. I asked the internet. https://www.ecenglish.com/learnenglish/lessons/prepositions-for-of-and-to

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java (hunk 10 above)
Comment (conniey) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: Ahaha. I got excited with those `<p>` tags. I am too used to lists and short statements. 🤣Oops.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClientBuilder.java [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+        policies.add(new ConfigurationCredentialsPolicy());
+        policies.add(new AsyncCredentialsPolicy(credentials));
+
+        policies.addAll(this.policies);
+
+        policies.add(new HttpLoggingPolicy(httpLogDetailLevel));
+
+        HttpPipeline pipeline = httpClient == null
+            ? new HttpPipeline(policies)
+            : new HttpPipeline(httpClient, policies);
```
[the remark this answers is not in the document]
Comment (conniey) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClientBuilder.java]: I wasn't sure if we should allow users to pass in an entire pipeline... I guess I'm worried that they could shoot themselves in the foot by not bothering to add in a credential policy and as a result, have something that doesn't work at all.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java (hunk 8 above)
Comment (JonathanGiles) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: The important question is: did you pass the test? :-)

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java (hunk 1 above)
Comment (JonathanGiles) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: I'm fine with whatever value is set (80, 100, 120, or whatever is fashionable these days)....I just wanted to make sure there was something telling us to wrap eventually! :-)

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClientBuilder.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (JonathanGiles) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClientBuilder.java]: I understand the reluctance. The exact phrase used in the arch board is "falling off the cliff". It's considered an escape hatch that might be useful in certain circumstances. I would recommend you consider making it settable, and make sure to update @g2vinay so he can follow an equivalent pattern in key vault.

[the remark this answers is not in the document]
Comment (conniey) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClientBuilder.java]: I see I see. 🤔
A couple of approaches could be to build the ConfigurationClient if they specify the pipeline since we don't need any other options.
```java
class ConfigurationClientBuilder {
    ConfigurationClient pipeline(HttpPipeline);
```
or
```java
class ConfigurationClientBuilder {
    ConfigurationClientBuilder pipeline(HttpPipeline);
```
And if the `pipeline` is set, to ignore all the other options and build the ConfigurationClient with that.

[the remark this answers is not in the document]
Comment (JonathanGiles) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClientBuilder.java]: For consistency I would go with the later approach, and document the precedence order.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java (hunk 1 above)
Comment (conniey) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: I think scarves are currently stylish. 🤣 120 works for me. I'll update it.

Comment (JonathanGiles) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: Fashion

Comment (conniey) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: https://github.com/Azure/azure-sdk-for-java/pull/3320

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java (hunk 5 above)
Comment (conniey) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: Fixed. I added documentation around this.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java (hunk 3 above)
Comment (conniey) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: Created an issue for this #3322 .

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java (hunk 6 above)
Comment (conniey) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: Created issue for this: #3323

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java (hunk 7 above)
Comment (conniey) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: image
Odd is that when I added the `</p>` and ran the javadoc commands, I get this error.

**Comment by conniey** [comment]
@JonathanGiles I think I've addressed all of your review comments... how does it look now?

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java (hunk 7 above)
Comment (JonathanGiles) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: odd. I wonder if it would be fixed if you put the closing paragraph at the end of the same line, rather than on a new line?

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/AzureConfiguration.java [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+    // User-Agent header value format: azsdk-<sdk_language>-<client_lib>/<sdk_version> <platform_info>
+    private static final String USER_AGENT_FORMAT = "azsdk-java-%s/%s %s";
+
+    //TODO: Eventually remove these hardcoded strings with https://github.com/Azure/azure-sdk-for-java/issues/3141
+    static final String NAME = "application-configuration";
+    static final String VERSION = "1.0.0-SNAPSHOT";
+
+    /*
+     * Gets the User-Agent header value using the provided SDK name and version.
+     * TODO: Move this out to the common library where it can be consumed by everyone
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/AzureConfiguration.java]: 👍

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java [hunk 15]
```diff
@@ -25,24 +18,23 @@
...
 
 /**
- * Client that contains all the operations for KeyValues in Azure Configuration Store.
+ * Client that contains all the operations for {@link ConfigurationSetting ConfigurationSettings} in Azure Configuration
+ * Store.
+ *
+ * @see ConfigurationClientBuilder
+ * @see ConfigurationClientCredentials
  */
 public final class ConfigurationClient extends ServiceClient {
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: Can you eagerly rename this class and the builder to `ConfigurationAsyncClient` and `ConfigurationAsyncClientBuilder`?

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java [hunk 16]
```diff
@@ -51,20 +43,33 @@
...
+        return new ConfigurationClientBuilder();
+    }
+
+    /**
+     * Adds a configuration value in the service if that key does not exist.
+     *
+     * @param key The key for the configuration setting to add.
+     * @param value The value associated with this configuration setting key.
+     * @return ConfigurationSetting that was created.
+     * @throws IllegalArgumentException If {@code key} or {@code value} are {@code null} or an empty string.
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: I'm guessing you're going to follow-up with the need to reduce client-side validation?

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java [hunk 17]
```diff
@@ -51,20 +43,33 @@
...
+    public static ConfigurationClientBuilder builder() {
+        return new ConfigurationClientBuilder();
+    }
+
+    /**
+     * Adds a configuration value in the service if that key does not exist.
+     *
+     * @param key The key for the configuration setting to add.
+     * @param value The value associated with this configuration setting key.
+     * @return ConfigurationSetting that was created.
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]:
```suggestion
     * @return The {@link ConfigurationSetting) that was created, or null, if a key collision occurs (which will also throw IllegalArgumentException described below).
```

Comment (JonathanGiles) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: Please apply a similar pattern as this for the other API to make it clear what is returned when the exception is thrown.

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java [hunk 18]
```diff
@@ -74,60 +79,89 @@ public static Builder builder() {
...
-        ConfigurationSetting result = validateSetting(setting);
-
+        ConfigurationSetting result = validateSetting(setting, true);
+
+        // This service method call is similar to addSetting except it will create or update a configuration setting.
+        // If the user provides an etag value, it is passed in as If-Match = "{etag value}". If the current value in the
+        // service has a matching etag then it matches, then its value is updated with what the user passed in.
+        // Otherwise, the service throws an exception because the current configuration value was updated and we have an
+        // old value locally.
+        // If no etag value was passed in, then the value is always added or updated.
```
Comment (JonathanGiles) [inline applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: 👍

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java (hunk 7 above)
Comment (conniey) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationClient.java]: `</p>` is not necessary when it is followed by `<ol>` and shows up as an error in the repo (on that mozilla link below) when I paste my javadoc in there. There are some other situations that don't require a `</p>`. I learned a lot.
https://developer.mozilla.org/en-US/docs/Web/HTML/Element/p

File: pom.client.xml [hunk 19]
```diff
@@ -162,18 +162,6 @@
...
-        <artifactId>azure-common</artifactId>
-        <version>${azure-common.version}</version>
-      </dependency>
-
-      <dependency>
-        <groupId>com.azure</groupId>
-        <artifactId>azure-common-auth</artifactId>
-        <version>${azure-common-auth.version}</version>
-      </dependency>
-
```
Comment (JonathanGiles) [inline pom.client.xml]: Not sure why this is removed?

Comment (conniey) [reply pom.client.xml]: It was added twice, so there was a warning. It still exists below here: https://github.com/Azure/azure-sdk-for-java/blob/db2c4ad69ef53eff0554fb8c51ad9dcdf1845fdf/pom.client.xml#L171-L181

Comment (JonathanGiles) [reply pom.client.xml]: 👍

File: applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java [hunk 20]
```diff
@@ -0,0 +1,343 @@
...
+ * @see ConfigurationClientCredentials
+ */
+public final class ConfigurationAsyncClient extends ServiceClient {
+    private static final String ETAG_ANY = "*";
+
+    private final String serviceEndpoint;
+    private final ConfigurationService service;
+
+    /**
+     * Creates a ConfigurationAsyncClient that uses {@code credentials} to authorize with Azure and {@code pipeline} to
```
[the remark this answers is not in the document]
Comment (conniey) [reply applicationconfig/client/src/main/java/com/azure/applicationconfig/ConfigurationAsyncClient.java]: I updated that.
